src/test/java/com/ning/http/client/async/PauseAsyncHandlerTest.java

```diff
@@ -0,0 +1,273 @@
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.ning.http.client.AsyncHandler;
+import com.ning.http.client.AsyncHttpClient;
+import com.ning.http.client.AsyncHttpClientConfig;
+import com.ning.http.client.HttpResponseBodyPart;
+import com.ning.http.client.HttpResponseHeaders;
+import com.ning.http.client.HttpResponseStatus;
+import com.ning.http.client.Response;
+import com.ning.http.client.providers.grizzly.PauseHandler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.handler.AbstractHandler;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public abstract class PauseAsyncHandlerTest extends AbstractBasicTest {
+
+    private static final int NUMBER_OF_CHUNKS = 5;
+    private static final String CHUNK_CONTENT = "This is a chunk";
+
+    public static class SlowChunkedHandler extends AbstractHandler {
+
+        public void handle(String pathInContext, Request request, HttpServletRequest httpRequest, HttpServletResponse httpResponse) throws IOException, ServletException {
+            httpResponse.setStatus(200);
+            httpResponse.setHeader("Transfer-encoding", "chunked");
+            httpResponse.setContentType("application/text");
+
+            httpResponse.flushBuffer();
+
+            final boolean wantFailure = httpRequest.getHeader("X-FAIL-TRANSFER") != null;
+            final boolean wantSlow = httpRequest.getHeader("X-SLOW") != null;
+
+            OutputStream os = httpResponse.getOutputStream();
+            for (int i = 0; i < NUMBER_OF_CHUNKS; i++) {
+                os.write(CHUNK_CONTENT.getBytes(), 0, CHUNK_CONTENT.length());
+
+                if (wantSlow) {
+                    try {
+                        Thread.sleep(300);
+                    } catch (InterruptedException ex) {
+                        // ignored
+                    }
+                }
+
+                if (wantFailure) {
+                    if (i == NUMBER_OF_CHUNKS - 1) {
+                        // Jetty will abort and drop the connection
+                        httpResponse.sendError(500);
+                        break;
+                    }
+                }
+
+                httpResponse.getOutputStream().flush();
+            }
+
+            httpResponse.getOutputStream().flush();
+            httpResponse.getOutputStream().close();
+        }
+    }
+
+    // a /dev/null but counting how many bytes it received
+    public static class CountingOutputStream extends OutputStream {
```

Comment:
same

Reply:
I extracted it instead, as it was copied from another test